### PR TITLE
Cleanup API

### PR DESCRIPTION
**Pull request by ppamorim**
I did some cleanup of the API to make it more concise and easy to maintain.

I also removed unnecessary conversion from `Data -> String -> JSON`, now the library converts `Data -> JSON` directly and the code got simplified.

This PR will be opened when #83 gets merged.

### CODE REVIEW

File: src/client.rs [hunk 1]
```diff
@@ -113,14 +118,22 @@ impl Client {
...
+                signed_key.update(request.as_bytes());
+                let signature = hex_encode(signed_key.finalize().into_bytes());
+                let request_body: String = format!("{:?}&signature={}", request, signature);
+                format!("{}{}?{}", self.host, String::from(endpoint), request_body)
+            },
+            None => {
+                let signed_key = Hmac::<Sha256>::new_varkey(self.secret_key.as_bytes()).unwrap();
+                let signature = hex_encode(signed_key.finalize().into_bytes());
+                let request_body: String = format!("&signature={}", signature);
+                format!("{}{}?{}", self.host, String::from(endpoint), request_body)
```
Comment (dorak88783) [inline src/client.rs:134]: This line is common in both arms of the match, you can pull it out

File: src/client.rs [hunk 2]
```diff
@@ -113,14 +118,22 @@ impl Client {
...
-        let signature = hex_encode(signed_key.finalize().into_bytes());
-        let request_body: String = format!("{}&signature={}", request, signature);
-        let url: String = format!("{}{}?{}", self.host, endpoint, request_body);
-
-        url
+    fn sign_request(&self, endpoint: API, request: Option<String>) -> String {
+        match request {
+            Some(request) => {
+                let mut signed_key = Hmac::<Sha256>::new_varkey(self.secret_key.as_bytes()).unwrap();
+                signed_key.update(request.as_bytes());
```
Comment (dorak88783) [inline src/client.rs:125]: I guess you can add this line to the previous, and then you can drop the `mut` on the `signed_key`. Not sure what is better Rust practice :)

**Review by dorak88783 — APPROVED** [review]

File: src/client.rs (hunk 1 above)
Comment (ppamorim) [reply src/client.rs:134]: They are different, on this one `signed_key.update(request.as_bytes())` is not called and the format from the `request_body` is different.

File: src/client.rs (hunk 2 above)
Comment (ppamorim) [reply src/client.rs:125]: Ok, we can discuss this later. Could you please review the #83 first?

File: src/client.rs (hunk 1 above)
Comment (dorak88783) [reply src/client.rs:134]: I was only referring to the last line which is common; the rest is indeed different. But it's a minor point.

**Comment by ppamorim** [comment]
@dorak88783 Missed to fix the general file.

**Comment by ppamorim** [comment]
@dorak88783 Could you please verify if all routes are correct?

**Comment by dorak88783** [comment]
@ppamorim I can try, but likely only next week. I was also working on creating end-to-end test cases using the mockito package, but need more time to get this into shape for a PR.

**Comment by dorak88783** [comment]
I had a quick scan and the endpoints look okay to me. Maybe you could sort them such that they appear in the order as in the API doc at https://github.com/binance/binance-spot-api-docs/blob/master/rest-api.md; and make a comment for the ones that are not implemented yet. I can also do this later in a new PR.

File: src/api.rs [hunk 3]
```diff
@@ -7,6 +7,93 @@ use crate::futures::market::*;
...
+    AllForceOrders,
+    OpenInterest,
+}
+
+impl From<API> for String {
+    fn from(item: API) -> Self {
+        match item {
+            API::Spot(route) => {
+                match route {
+                    Spot::Ping => String::from("/api/v3/ping"),
```
Comment (dorak88783) [inline src/api.rs:58]: I didn't try but you could pull out the `String::from(` at each line and put it at line 54, e.g. change line 54 into
`String::from( match item {`

**Comment by ppamorim** [comment]
@dorak88783 All done :)

**Comment by wisespace-io** [comment]
@ppamorim I pulled your branch and noticed some duplicated definitions. You can check the errors in https://travis-ci.org/github/wisespace-io/binance-rs ... also there are some clippy errors, you can also check in travis, or just run "cargo clippy -- -D warnings"

**Comment by ppamorim** [comment]
@wisespace-io This happened because draft PR #85 was merged before this PR. I create a PR to revert the changes. I also sorted the issues in this PR.

Please apply the revert and after review this PR.

**Comment by ppamorim** [comment]
@wisespace-io Build passed :) https://travis-ci.org/github/wisespace-io/binance-rs/builds/765038918

**Comment by dorak88783** [comment]
@wisespace-io : I'm honored to be asked to review :), but I don't think I'm a critical reviewer. I can have a look again but only on  Thursday; I think it's fine to merge like this.

**Comment by ppamorim** [comment]
@dorak88783 Any update?

**Comment by dorak88783** [comment]
Thanks for checking, but didn't have time today. Maybe in the weekend, but I'd not wait for me :)

**Review by dorak88783 — APPROVED** [review]

**Pull request merged**